### PR TITLE
[SPARK-12415] Do not use closure serializer to serialize task result

### PR DESCRIPTION
**Pull request by tedyu**
As the name suggests, closure serializer is for closures. We should be able to use the generic serializer for task results. If we want to do this we need to register `org.apache.spark.scheduler.TaskResult` if we use Kryo.

@andrewor14 
Please review

### CODE REVIEW

**Comment by tedyu** [comment]
```
ERROR: Timeout after 15 minutes
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/apache/spark.git
```

**Comment by tedyu** [comment]
Jenkins, test this please

**Comment by srowen** [comment]
Given the naming, I'm inclined to think you're right. CC @kayousterhout and @davies who have happened to touch these lines last.

**Comment by zsxwing** [comment]
Could you add a test for this issue? Such as, create a result that is not Serializable but supports `Kryo`.

**Comment by tedyu** [comment]
I took a look at ./core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala but may need some hint how such result should be formed.

**Comment by zsxwing** [comment]
Sorry, my comment was misleading. I'm wondering if you can write a simple test that uses `KryoSerializer `. Just to make sure we don't forget to register any class for `Kryo`.

**Comment by tedyu** [comment]
See if the new test makes sense

File: core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala [hunk 1]
```diff
@@ -81,6 +81,16 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
...
   def conf: SparkConf = new SparkConf().set("spark.akka.frameSize", "1")
 
+  test("Kryo serializer for TaskResult") {
+    val conf1 = new SparkConf(false)
+    conf1.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+    conf1.set("spark.kryoserializer.buffer", "1m")
+    conf1.set("spark.kryoserializer.buffer.max", "2m")
+    sc = new SparkContext("local", "test", conf1)
+    val result = sc.parallelize(Seq(1), 1).map(x => 2 * x).reduce((x, y) => x)
+    assert(result === 2)
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: can you explicitly create a `KryoSerializer` and use it to serialize both `TaskResult`s?

Comment (tedyu) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: Isn't the above action covered by the new test ?

Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: It's a more specific test. In case we didn't set this one up correctly for some reason this may still pass even if kryo doesn't actually work.

Comment (tedyu) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: Please take a look at createTaskResult() in core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala :
```
  def createTaskResult(id: Int): DirectTaskResult[Int] = {
    val valueSer = SparkEnv.get.serializer.newInstance()
    new DirectTaskResult[Int](valueSer.serialize(id), mutable.Map.empty, new TaskMetrics)
  }
```

Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: All I'm suggesting is a test that looks like the one added here:
https://github.com/apache/spark/commit/de0278286cf6db8df53b0b68918ea114f2c77f1f

In general tests should be as fine-grained as possible. Right now this just does an end-to-end test, which may be sufficient but is quite brittle. E.g. imagine if we change the name of the config or the serializer, then this won't actually test what we want.

**Comment by tedyu** [comment]
Modified new test in KryoSerializerSuite.scala

Looks like DirectTaskResult#equals should be defined to compare the fields.

Let me know if the above should be done.

**Comment by tedyu** [comment]
Jenkins, test this please

**Comment by tedyu** [comment]
Jenkins, test this please

**Comment by tedyu** [comment]
HeartbeatReceiverSuite.'expire dead hosts' failed

Doesn't seem to be related to the PR.

**Comment by tedyu** [comment]
Jenkins, test this please.

**Comment by tedyu** [comment]
@andrewor14 @zsxwing 
Please take another look.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 2]
```diff
@@ -215,6 +215,19 @@ class TaskMetrics extends Serializable {
...
+  override def equals(other: Any): Boolean = other match {
+    case that: TaskMetrics => {
+      if (this.executorDeserializeTime != that.executorDeserializeTime) return false
+      if (this.executorRunTime != that.executorRunTime) return false
+      if (this.resultSize != that.resultSize) return false
+      if (this.jvmGCTime != that.jvmGCTime) return false
+      if (this.resultSerializationTime != that.resultSerializationTime) return false
+      if (this.hostname == null && that.hostname != null) return false
+      if (this.hostname != null) this.hostname.equals(that.hostname) else true
+    }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: it would be better if this looked like:
```
this.executorDeserializeTime == that.executorDeserializeTime &&
this.executorRunTime == that.executorRunTime &&
...
this.hostname == that.hostname
```
no need to check `equals` here because we're in scala

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala [hunk 3]
```diff
@@ -46,6 +46,26 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
...
+      val thatAccumSize = if (that.accumUpdates != null) that.accumUpdates.size else 0
+      if (accumSize != thatAccumSize) return false
+      if (accumSize > 0) {
+        val b = this.accumUpdates.keys.forall { key =>
+          this.accumUpdates.get(key) == that.accumUpdates.get(key)
+        }
+        if (!b) return false;
+      }
+      this.metrics.equals(that.metrics)
+    }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: similarly, this can be simplified
```
val accumEquals =
  if (accumUpdates != null) {
    accumUpdates.keys.forall { k => accumUpdates.get(k) == that.accumUpdates.get(k) }
  } else {
    that.accumUpdates == null
  }
  valueBytes.equals(that.valueBytes) &&
  metrics.equals(that.metrics) &&
  accumEquals
```

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala [hunk 4]
```diff
@@ -109,6 +111,9 @@ class KryoSerializer(conf: SparkConf)
     kryo.register(classOf[SerializableJobConf], new KryoJavaSerializer())
     kryo.register(classOf[HttpBroadcast[_]], new KryoJavaSerializer())
     kryo.register(classOf[PythonBroadcast], new KryoJavaSerializer())
+    kryo.register(classOf[TaskMetrics], new KryoJavaSerializer())
+    kryo.register(classOf[DirectTaskResult[_]], new KryoJavaSerializer())
+    kryo.register(classOf[IndirectTaskResult[_]], new KryoJavaSerializer())
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: FYI, each of these calls is actually somewhat expensive. We create thousands of new Kryo serializer instances because we don't reuse them and so we end up calling `register` here many times. There is an outstanding issue [SPARK-12473](https://issues.apache.org/jira/browse/SPARK-12473) to resolve this, but until that is fixed this commit will likely cause minor performance regressions.

**Comment by andrewor14** [comment]
The rest of it looks OK. @zsxwing can you take another pass?

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala [hunk 5]
```diff
@@ -46,6 +46,26 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
...
+
+  override def equals(other: Any): Boolean = other match {
+    case that: DirectTaskResult[_] => {
+      if (!this.valueBytes.equals(that.valueBytes)) return false
+
+      val accumSize = if (accumUpdates != null) accumUpdates.size else 0
+      val thatAccumSize = if (that.accumUpdates != null) that.accumUpdates.size else 0
+      if (accumSize != thatAccumSize) return false
+      if (accumSize > 0) {
+        val b = this.accumUpdates.keys.forall { key =>
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: Why compare two Maps manually? I think `Map` already implements `equals`. Right?

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala (hunk 3 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: Also use `==` instead of `equals`.

File: core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala [hunk 6]
```diff
@@ -154,6 +155,19 @@ class KryoSerializerSuite extends SparkFunSuite with SharedSparkContext {
       mutable.HashMap(1 -> "one", 2 -> "two", 3 -> "three")))
   }
 
+  test("Bug: SPARK-12415") {
+    val ser = new KryoSerializer(conf.clone.set("spark.kryo.registrationRequired", "true"))
+      .newInstance()
+    def check[T: ClassTag](t: T) {
+      val ret = ser.deserialize[T](ser.serialize(t))
+      assert(ret.equals(t), "Deser " + ret + " orig " + t)
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/serializer/KryoSerializerSuite.scala]: Nit: Use `==` insead

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala [hunk 7]
```diff
@@ -46,6 +46,23 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
...
+    case that: DirectTaskResult[_] => {
+      val accumEquals =
+        if (accumUpdates != null) {
+          accumUpdates.keys.forall { k => accumUpdates.get(k) == that.accumUpdates.get(k) }
+        } else {
+          that.accumUpdates == null
+        }
+      valueBytes == that.valueBytes &&
+      metrics == that.metrics &&
+      accumEquals
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: We can just use `accumUpdates == that.accumUpdates` here.

File: core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala [hunk 8]
```diff
@@ -215,6 +215,18 @@ class TaskMetrics extends Serializable {
     inputMetrics.foreach(_.updateBytesRead())
   }
 
+  override def equals(other: Any): Boolean = other match {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: This is not a correct `equals`. There are other fields in `TaskMetrics`. To make `TaskMetrics.equals` work, there are also other classes need to be updated. Instead, could you just manually compare fields of DirectTaskResult in the tests rather than adding an incorrect `equals`?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala]: yeah, that would be preferable actually. In the future if we add more fields to `TaskMetrics` we might forget to add it to `equals`

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala [hunk 9]
```diff
@@ -46,6 +46,8 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
 
   def this() = this(null.asInstanceOf[ByteBuffer], null, null)
 
+  override def toString: String = valueBytes.toString + " " + accumUpdates + " " + metrics
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: nit: remove this

File: core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala [hunk 10]
```diff
@@ -220,7 +222,7 @@ class HeartbeatReceiverSuite
     } else {
       assert(!response.reregisterBlockManager)
       // Additionally verify that the scheduler callback is called with the correct parameters
-      verify(scheduler).executorHeartbeatReceived(
+      verify(scheduler, VerificationModeFactory.atLeast(1)).executorHeartbeatReceived(
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala]: nit: revert changes to `HeartbeatReceiverSuite`.

**Comment by zsxwing** [comment]
LGTM except some nits. Thanks a lot for fixing this @tedyu

File: core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala (hunk 9 above)
Comment (tedyu) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: This would make String form of TaskResult more readable.
Can this be kept ?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala]: This representation isn't really useful. It doesn't even tell you the task ID. I would just revert this change since it's unrelated and doesn't add much value.

**Comment by tedyu** [comment]
Here was why I implemented equals for Map (from #48257):
{code}
[info] - Bug: SPARK-12415 *** FAILED *** (3 milliseconds)
[info]   java.nio.HeapByteBuffer[pos=0 lim=2 cap=2] null org.apache.spark.executor.TaskMetrics@60d6ae17 did not equal java.nio.HeapByteBuffer[pos=0 lim=2 cap=2] Map() org.apache.spark.executor.TaskMetrics@2122833f Deser java.nio.HeapByteBuffer[pos=0 lim=2 cap=2] null org.apache.spark.executor.TaskMetrics@60d6ae17 orig java.nio.HeapByteBuffer[pos=0 lim=2 cap=2] Map() org.apache.spark.executor.TaskMetrics@2122833f (KryoSerializerSuite.scala:163)
{code}
One Map was null and one Map was empty.

**Comment by tedyu** [comment]
```
[info] PostgresIntegrationSuite:
[info] Exception encountered when attempting to run a suite with class name: org.apache.spark.sql.jdbc.PostgresIntegrationSuite *** ABORTED *** (9 seconds, 292 milliseconds)
[info]   com.spotify.docker.client.DockerRequestException: Request error: POST unix://localhost:80/containers/bc67f5d74a6cfb298d238cbc53eb6540564c4e642ac630e6448da4dd0871fc37/start: 500
[info]   at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1133)
[info]   at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1104)
[info]   at com.spotify.docker.client.DefaultDockerClient.startContainer(DefaultDockerClient.java:401)
```
Don't think the above was related to this PR

**Comment by tedyu** [comment]
Jenkins, test this please

**Comment by tedyu** [comment]
Jenkins, test this please

**Comment by tedyu** [comment]
```
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 209 times over 10.019028894 seconds. Last failure message: strings.forall({
  ((elem: Any) => DirectKafkaStreamSuite.collectedData.contains(elem))
}) was false.
	at org.scalatest.concurrent.Eventually$class.tryTryAgain$1(Eventually.scala:420)
	at org.scalatest.concurrent.Eventually$class.eventually(Eventually.scala:438)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite.eventually(DirectKafkaStreamSuite.scala:44)
	at org.scalatest.concurrent.Eventually$class.eventually(Eventually.scala:307)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite.eventually(DirectKafkaStreamSuite.scala:44)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite$$anonfun$5.org$apache$spark$streaming$kafka$DirectKafkaStreamSuite$$anonfun$$sendDataAndWaitForReceive$1(DirectKafkaStreamSuite.scala:249)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite$$anonfun$5.apply$mcV$sp(DirectKafkaStreamSuite.scala:310)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite$$anonfun$5.apply(DirectKafkaStreamSuite.scala:235)
	at org.apache.spark.streaming.kafka.DirectKafkaStreamSuite$$anonfun$5.apply(DirectKafkaStreamSuite.scala:235)
```
Not sure if the test failure in build #48288 was related.

File: core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala (hunk 10 above)
Comment (srowen) [inline core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala]: @tedyu this still needs attention. Is the rest OK to commit @andrewor14 or are you interested in refining the tests?

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala (hunk 4 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: I just found there are more classes need to register, such as `ShuffleReadMetrics`, `OutputMetrics` and `ShuffleWriteMetrics`. The problem is that most fields in `TaskMetrics` are `Option` so the new added unit tests cannot cover them.

Not sure how many people uses `spark.kryo.registrationRequired`. But if they enable it, we will break their codes.

In addition, even if we register all necessary classes now, people may forget to register new classes if they just add an `Option` field in future.

I prefer to close this PR since I don't think this patch buys us much due to the tricky thing introduced.

Comment (tedyu) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: bq. people may forget to register new classes if they just add an Option field to TaskMetrics in future

Addition to TaskMetrics would be reviewed, right ?
A comment can be added to TaskMetrics reminding them to register corresponding class.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: Moreover, `_accumulatorUpdates` is much more complicated. Its value could be any type, e.g., `SQLMetricValue`. If someone adds a new `SQLMetricValue` type, it's hard for reviewers to recall it will be stored in `TaskMetrics`.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: I see, that does seem like a real problem. Let's move this discussion to the main thread so more people can see.

**Comment by andrewor14** [comment]
As @zsxwing pointed out, the existing unit tests are not sufficient because they didn't catch the fact that `ShuffleReadMetrics` et al must also be registered. This is because these fields are options. In the future, people are free to add more options fields and it will be hard for reviewers to maintain this additional requirement of registering it with Kryo.

Given this the suggestion is to close this issue for now. Eventually when we get rid of closure serializer altogether (SPARK-12414) we can just replace this one with Java serializer, and add a huge comment why it can't be Kryo.

File: core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala [hunk 11]
```diff
@@ -81,6 +81,16 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
   // as we can make it) so the tests don't take too long.
   def conf: SparkConf = new SparkConf().set("spark.akka.frameSize", "1")
 
+  test("Kryo serializer for TaskResult") {
+    val conf1 = new SparkConf(false)
+    conf1.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
```
Comment (tedyu) [inline core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala]: I wasn't aware of SPARK-12414 previously.
If Java serializer is to be used everywhere, is there some code in this PR that is useful for SPARK-12414 ?

**Comment by tedyu** [comment]
I think the first two days' work aligns with SPARK-12414.

Is it possible to keep that (without modification to KryoSerializer and the new test) ?

**Comment by andrewor14** [comment]
What first two days of work are you referring to? The problem is we just can't use Kryo for serializing task results because there might be unregistered classes. Because of this constraint we can't use `spark.serializer` here since the user can specify Kryo there.

**Comment by andrewor14** [comment]
Thanks for putting the effort into this @tedyu. It did move us forward in the sense that now we know we can't use Kryo for serializing task results. I'm going to close the issue.

**Comment by kervin521** [comment]
Java：org.apache.spark.SparkException: Task not serializable

Any method used in relation to sorting will be wrongly reported “Task not serializable”

EX:

sims.top(K, new Comparator<Tuple2<Object, Double>>(){})

sc.parallelize(Arrays.asList(_data)).zipWithIndex().sortByKey(new Comparator<Double>)
